samcli/lib/list/endpoints/endpoints_producer.py

```diff
@@ -0,0 +1,483 @@
+                return []
+            LOG.error("ClientError Exception : %s", str(e))
+            raise SamListUnknownClientError(msg=str(e)) from e
+        except BotoCoreError as e:
+            LOG.error("Botocore Exception : %s", str(e))
+            raise SamListUnknownBotoCoreError(msg=str(e)) from e
+
+    def build_api_gw_endpoints(self, physical_id: str, stages: list) -> list:
+        """
+        Builds the default api gateway endpoints
+
+        Parameters
+        ----------
+        physical_id: str
+            The physical ID of the api resource
+        stages: list
+            A list of stages for the api resource
+
+        Returns
+        -------
+        api_list: List[Any]
+            The list of default api gateway endpoints
+        """
+        api_list = []
+        for stage in stages:
+
+            api_list.append(f"https://{physical_id}.execute-api.{self.region}.amazonaws.com/{stage}")
+        return api_list
+
+    def get_api_gateway_endpoint(
+        self, deployed_resource: Dict[Any, Any], custom_domain_substitute_dict: Dict[Any, Any]
+    ) -> Any:
+        """
+        Gets the API gateway endpoints for APIGateway and APIGatewayV2 APIs
+
+        Parameters
+        ----------
+        deployed_resource: Dict[Any, Any]
+            Dictionary containing the resource info of the deployed API
+        custom_domain_substitute_dict: Dict[Any, Any]
+            Dictionary containing the mappings of the custom domains for APIs
+
+        Returns
+        -------
+        endpoint: Any
+            The endpoint(s) of the current API resource
+        """
+        endpoint: Any
+        stages = self.get_stage_list(
+            deployed_resource.get(PHYSICAL_RESOURCE_ID, ""),
+            get_api_type_enum(deployed_resource.get(RESOURCE_TYPE, "")),
+        )
+        if deployed_resource.get(LOGICAL_RESOURCE_ID, "") in custom_domain_substitute_dict:
+            endpoint = custom_domain_substitute_dict.get(deployed_resource.get(LOGICAL_RESOURCE_ID, ""), "-")
+        else:
+            endpoint = self.build_api_gw_endpoints(deployed_resource.get(PHYSICAL_RESOURCE_ID, ""), stages)
+        return endpoint
+
+    def get_cloud_endpoints(self, stacks: list) -> list:
+        """
+        Gets a list of cloud endpoints resources
+
+        Parameters
+        ----------
+        stacks: list
+            A list containing the local stack
+
+        Returns
+        -------
+        endpoints_list: List[Any]
+            A list of cloud endpoints resources
+        """
+        endpoints_list = []
+        local_stack = stacks[0]
+        local_stack_resources = local_stack.resources
+        seen_endpoints = set()
+        response = self.get_resources_info()
+        response_domain_dict = get_response_domain_dict(response)
+        custom_domain_substitute_dict = get_custom_domain_substitute_list(response, stacks, response_domain_dict)
+
+        for deployed_resource in response.get(STACK_RESOURCES, {}):
```

Comment:
Same comment as I had for other one before (apologies that I missed this). I know we have some tests to validate this but it is too overloaded when someone takes a look at this for loop for the first time. Comments or more logging might help us to deep dive into an issue.